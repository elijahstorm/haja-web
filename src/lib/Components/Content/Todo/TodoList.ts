import { storeQuery } from "$lib/firebase/firestore"
import type { DocumentData, QuerySnapshot } from "firebase/firestore"
import type { TodoContentConfig } from "./TodoContent"

export const getTodoList: (input: {
	source: string
	dateRange?: { start: Date; end: Date }
	isTeam?: boolean
	amount?: number
}) => Promise<TodoContentConfig[] | string> = async ({
	source,
	dateRange = { start: new Date(0), end: new Date() },
	isTeam = false,
	amount = 50
}) => {
	let list: QuerySnapshot<DocumentData>

	try {
		list = await storeQuery({
			type: "todo",
			source,
			amount,
			isTeam,
			timestamp: "date",
			queries: [
				{
					type: "date",
					compare: ">=",
					value: dateRange.start
				},
				{
					type: "date",
					compare: "<",
					value: dateRange.end
				}
			]
		})
	} catch (e) {
		return e
	}

	return list.docs.map((doc) => {
		const data = doc.data()

		return {
			contentType: "todo",
			id: data.id ?? doc.id,
			title: data.title?.trim(),
			caption: data.caption?.trim(),
			color:
				!data.color || data.color === ""
					? "var(--primary)"
					: `#${data.color.substring(2) + data.color.substring(0, 2)}`,
			date: data.date.toDate(),
			status: data.status,
			type: data.type
		}
	})
}
