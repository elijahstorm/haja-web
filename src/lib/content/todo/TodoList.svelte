<script lang="ts">
	import { base } from "$app/paths"
	import AddTodo from "./AddTodo.svelte"
	import type { TodoContentConfig } from "./TodoContent"
	import TodoContent from "./TodoContent.svelte"

	export let todos: TodoContentConfig[] | string = []
	export let source: string
	export let isTeam: boolean
	export let locked: boolean = false

	let push = false
	let page = 1
	let list = todos

	const api = `${base}/api/todos&source=${source}&isTeam=${isTeam}`

	function infiniteHandler({ detail: { loaded, complete } }) {
		fetch(`${api}&page=${page}`)
			.then((response) => response.json())
			.then((data) => {
				if (data.hits.length) {
					page += 1
					list = [...list, ...data.hits]
					loaded()
				} else {
					complete()
				}
			})
	}

	const callback = (todo: TodoContentConfig) => {
		if (typeof todos === "string") return

		todos = [todo, ...todos]
		push = true

		setTimeout(() => {
			push = false
		}, 1100)
	}
</script>

{#if !locked}
	<div class="add" class:push>
		<AddTodo {source} {isTeam} {callback} />
	</div>
{/if}
{#if typeof todos === "string" || !Array.isArray(todos)}
	{todos}
{:else if todos.length == 0}
	<p>No todos yet!</p>
{:else}
	<div class:push>
		{#each todos as todo (todo.id)}
			<div class="todo">
				<TodoContent {todo} {source} {isTeam} />
			</div>
		{/each}
	</div>
{/if}

<style>
	.todo {
		margin: 1rem 0;
	}
	:not(.add).push {
		transform: translateY(-3rem);
		animation-name: push;
		animation-delay: 100ms;
		animation-duration: 1000ms;
		animation-timing-function: ease;
	}
	.add.push {
		opacity: 0;
		animation-name: fade;
		animation-delay: 600ms;
		animation-duration: 500ms;
		animation-timing-function: ease;
	}
	p {
		margin-top: 2rem;
	}

	@keyframes push {
		from {
			transform: translateY(-3rem);
		}
		to {
			transform: translateY(0);
		}
	}
	@keyframes fade {
		from {
			opacity: 0;
		}
		to {
			opacity: 1;
		}
	}
</style>
