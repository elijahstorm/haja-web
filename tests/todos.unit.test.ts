import { describe, it, expect, afterEach } from "vitest"
import { cleanup, fireEvent, render } from "@testing-library/svelte"
import TodoList from "$lib/Components/Content/Todo/TodoList.svelte"
import type { TodoContentConfig } from "$lib/Components/Content/Todo/TodoContent"

describe("TodoList.svelte", () => {
	afterEach(() => cleanup())

	const source = "source"
	const todos: TodoContentConfig[] = [
		{
			id: "1",
			contentType: "todo",
			date: new Date(),
			title: "title 1",
			caption: "caption 1",
		},
		{
			id: "2",
			contentType: "todo",
			date: new Date(),
			title: "title 2",
			caption: "caption 2",
		},
	]

	it("can add more if unlocked", () => {
		const { container } = render(TodoList, { todos, source, isTeam: false, locked: false })
		expect(container).toBeTruthy()
		expect(container.innerHTML).toContain("Add a new todo")
		expect(container.innerHTML).toContain("title 1")
		expect(container.innerHTML).toContain("caption 2")
	})

	it("cannot add more if locked", () => {
		const { container } = render(TodoList, { todos, source, isTeam: false, locked: true })
		expect(container).toBeTruthy()
		expect(container.innerHTML).not.toContain("Add a new todo")
		expect(container.innerHTML).toContain("title 1")
		expect(container.innerHTML).toContain("caption 2")
	})

	it("shows empty todo list", () => {
		const { container } = render(TodoList, { todos: [], source, isTeam: false, locked: true })
		expect(container).toBeTruthy()
		expect(container.innerHTML).toContain("No todos")
	})

	it("can add more if unlocked", async () => {
		const { container } = render(TodoList, { todos, source, isTeam: false, locked: false })
		expect(container).toBeTruthy()
		expect(container.innerHTML).toContain("Add a new todo")
		await fireEvent.click(container.querySelector('[data-test-type="add"]'))
		expect(container.innerHTML).toContain("Let's do Together!")
	})
})
