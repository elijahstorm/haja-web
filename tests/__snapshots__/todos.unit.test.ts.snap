// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList.svelte > can add more if unlocked 1`] = `"<div><div class=\\"add s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 s-AKPngUPfGbRU cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><div class=\\"s-AKPngUPfGbRU\\"><!--<Icon>--></div> <div class=\\"text-lg my-auto mx-0 s-AKPngUPfGbRU\\"><span class=\\"font-bold s-AKPngUPfGbRU\\">Add a new todo</span></div></div><!--<AddTodo>--></div> <div class=\\"s-HZ-oYOGGXvyC\\"><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 1</p> <p class=\\"opacity-60\\">caption 1</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 2</p> <p class=\\"opacity-60\\">caption 2</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div></div><!--<TodoList>--></div>"`;

exports[`TodoList.svelte > can add more if unlocked 2`] = `"<div><div class=\\"add s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 s-AKPngUPfGbRU\\" data-test-type=\\"add\\" role=\\"button\\" tabindex=\\"0\\"><div class=\\"s-AKPngUPfGbRU\\"><!--<Icon>--></div> <input class=\\"font-bold w-full border-none text-lg background-transparent s-AKPngUPfGbRU\\" placeholder=\\"Let's do Together!\\"></div><!--<AddTodo>--></div> <div class=\\"s-HZ-oYOGGXvyC\\"><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 1</p> <p class=\\"opacity-60\\">caption 1</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 2</p> <p class=\\"opacity-60\\">caption 2</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div></div><!--<TodoList>--></div>"`;

exports[`TodoList.svelte > cannot add more if locked 1`] = `"<div> <div class=\\"s-HZ-oYOGGXvyC\\"><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 1</p> <p class=\\"opacity-60\\">caption 1</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div><div class=\\"my-4 mx-0 s-HZ-oYOGGXvyC\\"><div class=\\"flex gap-4 items-start\\"><div class=\\"relative cursor-pointer\\" role=\\"button\\" tabindex=\\"0\\"><!--<Icon>--></div> <div class=\\"flex flex-col gap-3\\"><div class=\\"cursor-pointer text-lg my-auto mx-0 flex flex-col\\" role=\\"button\\" tabindex=\\"0\\"><p class=\\"font-bold\\">title 2</p> <p class=\\"opacity-60\\">caption 2</p> </div> <!--<InlineTodoEditor>--></div></div><!--<TodoContent>--> </div></div><!--<TodoList>--></div>"`;

exports[`TodoList.svelte > shows empty todo ist 1`] = `"<div></div><div></div><div> <p class=\\"mt-8 s-HZ-oYOGGXvyC\\">No todos yet!</p><!--<TodoList>--></div>"`;
